### PR TITLE
[Security] $attributes can be anything, but RoleVoter assumes strings

### PR DESCRIPTION
**Pull request by maennchen**
| Q | A |
| --- | --- |
| Branch? | 2.7 |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | yes |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #18042 |
| License | MIT |
| Doc PR | reference to the documentation PR, if any |

### CODE REVIEW

**Comment by iltar** [comment]
Considering this has most likely never worked, I think attributes should be hinted as string instead.

**Comment by maennchen** [comment]
@iltar I don't think that this is a good idea.

I have a use case where i want to check if some operations (attributes) can be applied on an object (subject). 

For sure those operations could be flattened into a string, but I'd have to parse them in the voter which is total nonsense.

Type Hinting for string would be a BC break since I can supply objects as attributes already, but only if they provide a `__toString()` method.

**Comment by iltar** [comment]
I guess implementing `__toString()` is already a debatable feature on its own. For the other cases, you have examples of how this would be used? I can't imagine because it's not how it's documented.

**Comment by Gladhon** [comment]
@iltar i think its basically a good idea to allow objects there, cause in that case you can easy check if your voter is responsible or not. Like ExpressionVoter: https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Security/Core/Authorization/Voter/ExpressionVoter.php#L60

Of course you also can do that with strings, and prefix strings, and hope that there are no conflicts ... but who want this?

**Comment by maennchen** [comment]
@iltar I only implemented the `__toString` to maintain backwards compatibility. In my opinion I'd rather remove that.

But there are already examples in Symfony itself using non string values. (See @Gladhon's comment) The `ExpressionVoter` only works because the `Expression` implements `__toString`.

**Comment by hhamon** [comment]
Otherwise you must implement your custom voter that support your custom type of attribute.

**Comment by hhamon** [comment]
Btw, the border between bug and new feature is really close in this case. I think it's better to consider it a new feature and submit the PR against the `master` branch.

**Comment by maennchen** [comment]
@hhamon I'd like your proposed solution. I implemented the `__toString()` thing to avoid BC breaks.

**Comment by maennchen** [comment]
I think that it's definitely a bug since the interface clearly says that mixed is accepted. (Which is currently only if you implement `__toString`.
Could we implement this as a bug and introduce a new feature for master which cleans the behaviour up?

**Comment by sstok** [comment]
`supportsAttribute` for RoleVoter should only allow a string or Role(Interface) (`Role` is actually the only one that is officially supported, because of the string promise).

But when the attributes of a not supported type, the de decision should be abstained?
When passing an unsupported attribute it would have failed with an php error, no?

**Comment by maennchen** [comment]
@sstok Two times yes.

**Comment by maennchen** [comment]
Also `supportsAttribute` does not exist anymore in `>= 3.0` I think.

**Comment by hhamon** [comment]
Yes `supportAttributes`  as a "public" method was removed in Symfony 3.0 because this method was never called by the Security component in Symfony 2.

**Comment by maennchen** [comment]
Should I change to check if the attribute is instance of RoleInterface instead of __toString? This will be a small BC break.

**Comment by ro0NL** [comment]
@maennchen why exactly do you want to remove `__toString` support? Imo we should first check for `RoleInterface` then string/string-object.

That would still be a BC break, but personally i consider it a bugfix.

File: src/Symfony/Component/Security/Core/Authorization/Voter/RoleVoter.php [hunk 1]
```diff
@@ -37,6 +37,13 @@ public function __construct($prefix = 'ROLE_')
...
     public function supportsAttribute($attribute)
     {
+        if (is_array($attribute)) {
+            return false;
+        }
+        if (is_object($attribute) && !method_exists($attribute, '__toString')) {
+            return false;
+        }
+
         return 0 === strpos($attribute, $this->prefix);
```
Comment (ro0NL) [inline src/Symfony/Component/Security/Core/Authorization/Voter/RoleVoter.php]: Could be simply `(string) $attribtute` as well. No need for the above statements... it should crash for invalid values by design.

Comment (maennchen) [inline src/Symfony/Component/Security/Core/Authorization/Voter/RoleVoter.php]: An attribute is mixed by design. It shouldn't crash for non string values. It should abstain. Why should it be a convention that every object passed as an attribute can be converted to a string.

I implemented a Voter, that was designed to vote for Operations on a subject when I encountered the error. Why should those operations which are not by any means strings have a toString method? That makes no sense for something that is by design mixed.

(By the way: strpos will automatically do a string casting. No string cast needed for that.)

Comment (maennchen) [inline src/Symfony/Component/Security/Core/Authorization/Voter/RoleVoter.php]: I'd change the code to check if `$attribute instanceof RoleInterface`, if yes do a `getRole()` on it. If it is no string or RoleInterface, I'd continue / ABSTAIN.

Comment (ro0NL) [inline src/Symfony/Component/Security/Core/Authorization/Voter/RoleVoter.php]: > It shouldn't crash for non string values. It should abstain

Sorry, you're right. However i think for 2.7 `supportAttribute` should also consider `RoleInterface` due mixed attribute support.

**Comment by maennchen** [comment]
I refactored it with specific checks for `RoleInterface`.

I personally don't really care if there is support for objects having a `__toString()` method. In my opinion it's cleaner just to handle strings & `RoleInterface`, but with `__toString()` support there is no BC break.

File: src/Symfony/Component/Security/Core/Authorization/Voter/RoleVoter.php [hunk 2]
```diff
@@ -57,6 +58,10 @@ public function vote(TokenInterface $token, $object, array $attributes)
         $roles = $this->extractRoles($token);
 
         foreach ($attributes as $attribute) {
+            if ($attribute instanceof RoleInterface) {
+                $attribute = $attribute->getRole();
```
Comment (fabpot) [inline src/Symfony/Component/Security/Core/Authorization/Voter/RoleVoter.php]: We should not call `supportAttribute` if the return value of `getRole()` is `null` as it means that there is no good representation of the role as a string.

Comment (fabpot) [inline src/Symfony/Component/Security/Core/Authorization/Voter/RoleVoter.php]: nevermind, that's already covered in `supportsAttribute()` via the `is_string()` check.

**Comment by fabpot** [comment]
Thank you @maennchen.

**Pull request merged**